File: Java/GenerateParenthesis.java
class Solution {

    List<String> res = new ArrayList<>();

    private void generate(String s, int left, int right) {
        if(left == 0 && right == 0) res.add(s);
        if(left > 0) generate(s + "(", left - 1, right);
        if(right > left) generate(s + ")", left, right - 1);
    }

    public List<String> generateParenthesis(int n) {
        generate("", n, n);
        return res;
    }
}

File: Go/generate_parenthesis.go
func generate(res *[]string, s []byte, left, right int) {
    if left == 0 && right == 0 {
        *res = append(*res, string(s))
    }
    if left > 0 {
        generate(res, append(s, '('), left - 1, right)
    }
    if right > left {
        generate(res, append(s, ')'), left, right - 1)
    }
}

func generateParenthesis(n int) []string {
    res := make([]string, 0)
    var s []byte
    generate(&res, s, n, n)

    return res
}